docs/ref/csp.txt

```diff
@@ -0,0 +1,279 @@
+
+Some key benefits of implementing CSP include:
+
+1. Mitigating XSS attacks by preventing the execution of inline scripts and restricting which external
+   scripts can be loaded.
+2. Controlling which external resources (like images, fonts, or stylesheets) can be loaded by your site.
+3. Preventing unwanted framing of your site, protecting against clickjacking attacks.
+4. Reporting policy violations to a specified URL, allowing you to monitor potential attacks or
+   misconfigured directives.
+
+CSP provides a robust and flexible framework for securing your web application, giving you fine-grained
+control over the resources your site can use and interact with. By leveraging Django's CSP middleware,
+you can easily implement and manage these policies in your project.
+
+Using CSP
+=========
+
+To enable CSP in your Django project:
+
+1. Add the CSP middleware to your ``MIDDLEWARE`` setting:
+
+   .. code-block:: python
+
+      MIDDLEWARE = [
+          # ...
+          "django.middleware.csp.ContentSecurityPolicyMiddleware",
+          # ...
+      ]
+
+2. Configure your CSP policies in your ``settings.py`` using either ``SECURE_CSP`` or ``SECURE_CSP_REPORT_ONLY``:
+
+   .. code-block:: python
+
+      from django.middleware import csp
+
+      SECURE_CSP = {
+          "DIRECTIVES": {
+              "default-src": [csp.SELF],
+              "script-src": [csp.SELF],
+              "style-src": [csp.SELF],
+              # Add more directives as needed
+          }
+      }
+
+      # Or for report-only mode:
+      SECURE_CSP_REPORT_ONLY = {
+          "DIRECTIVES": {
+              "default-src": [csp.SELF],
+              "script-src": [csp.SELF],
+              "style-src": [csp.SELF],
+              # Add more directives as needed
+          }
+      }
+
+3. To use CSP nonces for inline scripts or styles, include ``csp.NONCE`` in your directives:
+
+   .. code-block:: python
+
+      SECURE_CSP = {
+          "DIRECTIVES": {
+              "script-src": [csp.SELF, csp.NONCE],
+              # Other directives...
+          }
+      }
+
+Nonce Usage
+===========
+
+To use nonces in your CSP policy:
+
+1. Add the CSP context processor to your ``TEMPLATES`` setting:
+
+   .. code-block:: python
+
+      TEMPLATES = [
+          {
+              "BACKEND": "django.template.backends.django.DjangoTemplates",
+              "OPTIONS": {
+                  "context_processors": [
+                      # ...
+                      "django.template.context_processors.csp_nonce",
+                  ],
+              },
+          },
+      ]
+
+   This context processor makes the CSP nonce available in all templates.
+
+2. In your templates, use the ``request.csp_nonce``:
+
+   .. code-block:: html
+
+      <script nonce="{{ request.csp_nonce }}">
+          // Your inline JavaScript
+      </script>
+
+The CSP middleware will automatically generate and include a nonce in the CSP header for each request.
```

Comment:
Suggestion: My reading of the middleware is that the nonce is only added if the nonce value is read at least once during the request. It might be worth calling that out to aid the "I've set up a nonce but it's not in the header" debugging.

Reply:
Very true. I'll add a note about that. Thanks for pointing it out. 

I'll also add that if this is something we don't want to do in Django, it could be changed. I think it's helpful that if the nonce is never referenced it doesn't get added to the header, but it can also be a source of confusion. It could also be a source of confusion if the nonce is in header and not the content. But perhaps less so. I won't change anything now but just offering it up as open to discussion.

Reply:
I also believe the implementation is that if the nonce is read, but not included in settings, it'll be added to the `script` tag, but not the header. In those cases, I assume the nonce is just ignored and no load restriction is put in. That could also be a potential source of confusion.

Reply:
Added the note below.